### ISSUE TITLE
Non-string values in messages break filters if used

### ISSUE THREAD

**Issue by jmichalicek** [andrewgodwin/channels#274]
If you have a consumer which looks something like
```
def my_consumer(message):
    object_id = message.get('object_id')
    my_object = MyObject.objects.get(id=object_id)
    # do more stuff...
```

With a route of
```
route('my_channel', my_consumer)
```

You can send a message which has non-string values without any issue and they get serialized, assuming they are serializable of course, but for now I'm just looking at simple things like integers which would be a very common case for passing django object ids around.
```
# my_consumer will get this and the object_id value will be an int
# assuming some_object.id is an int as with standard django stuff
Channel('my_channel').send({'object_id': some_object.id})
````

But if you change that route to
```
route('my_channel', my_consumer, object_id=r'^(?P<object_id>\d+)')
```
And consumer to
```
def my_consumer(message, object_id):
    my_object = MyObject.objects.get(id=object_id)
    # do more stuff... use the orm to look up the object, etc.
```

Then the routing raises `TypeError: expected string or buffer` from `Route.match()` at https://github.com/andrewgodwin/channels/blob/master/channels/routing.py#L152 due to trying to match the regex against the int.

Does it make sense to have `Router.normalise_re_arg()` special case types other than six.binary_type which are easily handleable and readily serializable such as int, float, maybe others such as Decimal?  Or is it intended that message keys used in filters will always have string values?

**Comment by andrewgodwin** [comment]
I'm not sure I want to encourage regex matching against non-string types - the need for a canonical form means that `float` is not going to go down well (and `Decimal` cannot be transported in ASGI so that's not needed).

It looks like you're just trying to capture it to use as a kwarg - is there a reason you wish to do this rather than just use `message['object_id']` inside the body of the function?

**Comment by jmichalicek** [comment]
It felt like that was part of the intended use of being able to use those filters and captures in the routes and makes behavior for the same key consistent, whether it's accessed via the message dict or as an actual function argument captured from the filters.

I totally get it if the intent is that any keys you use for filters are intended to have string values, though.  There are always going to be values which would serialize correctly for use from the message dict but have no sane way of normalizing them for the filters regex and even simple values like floats can get into the occasional questionable area.  So it may be easiest/safest to just officially say "only use strings for this use case".

**Comment by andrewgodwin** [comment]
So, thinking about this a bit more, two possibilities I'm not sure about yet:

* Integers can be used in filters and must be compared directly to integers
* You can pass arbitrary functions instead of just regexes and that lets you use `lambda x: 50 < x < 400`

Neither of these would let you capture groups, though, like you were trying, as that's purely a function of regexes and strings and I don't think it makes sense on other types.

**Comment by jmichalicek** [comment]
Apparently I got sidetracked and never clicked the comment button...

Well, that second option definitely has some useful potential even if it does not allow the `x` to be assigned back to the arg for being passed into the consumer function.

The first option probably has some limited use, and I'm generally all for giving people the choice to do something in case it is the right solution, but probably isn't something I'd tend towards using as it looks like it could spiral out of control in the routes and is covered by just having `my_route_1`, `my_route_2` and some code like `Channel('my_route_%d' % some_int).send({'foo': 'bar'})` if someone really does want that.  I suppose arguably it would be cleaner than fiddling with the channel name string in that it would be sending to the same channel name every time with a different value for the param passed in.

This is definitely not a "must have" for me.  I was just checking on that in case it was intended to work how I was trying to use it before I went about just using the dict.  This is an awesome system and once it's in django core, it's less to keep up to date vs using celery and gunicorn and gives me websockets on top of that..
